packages/opentelemetry-node-tracer/src/instrumentation/PluginLoader.ts

```diff
@@ -0,0 +1,146 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Logger, Plugin, Tracer } from '@opentelemetry/types';
+import * as hook from 'require-in-the-middle';
+import * as utils from './utils';
+
+// States for the Plugin Loader
+export enum HookState {
+  UNINITIALIZED,
+  ENABLED,
+  DISABLED,
+}
+
+interface PluginNames {
+  [pluginName: string]: string;
+}
+
+// tslint:disable-next-line:no-any
+export type PluginType = Array<Plugin<any>>;
+
+/**
+ * The PluginLoader class can load instrumentation plugins that
+ * use a patch mechanism to enable automatic tracing for
+ * specific target modules.
+ */
+export class PluginLoader {
+  /** A list of loaded plugins. */
+  private _plugins: PluginType = [];
+  /**
+   * A field that tracks whether the require-in-the-middle hook has been loaded
+   * for the first time, as well as whether the hook body is activated or not.
+   */
+  private _hookState = HookState.UNINITIALIZED;
+
+  /** Constructs a new PluginLoader instance. */
+  constructor(readonly tracer: Tracer, readonly logger: Logger) {}
+
+  get state(): HookState {
+    return this._hookState;
+  }
+
+  get plugins(): PluginType {
+    return this._plugins;
+  }
+
+  /**
+   * Loads a list of plugins. Each plugin module should implement the core
+   * {@link Plugin} interface and export an instance named as 'plugin'. This
+   * function will attach a hook to be called the first time the module is
+   * loaded.
+   * @param modulesToPatch A list of plugins.
+   */
+  load(modulesToPatch: string[]): PluginLoader {
```

Comment:
It would be nice to add options to pass for each module to patch e.g
pluginLoader.load(['http'], { http: { httpTimings: true, filterings: ['/healthcheck'] } });

Reply:
Added in `Plugin` interface PR: https://github.com/open-telemetry/opentelemetry-js/pull/110/files#diff-0e705bb0dee9becebae81dd8b49650fdR31, ptal. An optional config an object to configure the plugin.

Reply:
I still think enabling and configuring should be done in a single configuration. This means that instead of accepting an array, the loader should accept an object that for now would accept only booleans but can later be extended with configuration options. It will also make it easier to understand that you are not completely overriding the default by doing this.

Reply:
SGTM, changed to an object (`PluginConfig`) instead of an array. Also, added `@todo` to add support for configuration options in the future. PTAL.